datafusion/optimizer/src/simplify_expressions/inlist_simplifier.rs

```diff
@@ -0,0 +1,106 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! This module implements a rule that simplifies expressions that is guaranteed to be true or false at planning time
+
+use std::collections::HashSet;
+
+use datafusion_common::tree_node::TreeNodeRewriter;
+use datafusion_common::Result;
+use datafusion_expr::expr::InList;
+use datafusion_expr::{lit, BinaryExpr, Expr, Operator};
+
+/// Simplify expressions that is guaranteed to be true or false to a literal boolean expression
+///
+/// Rules:
+/// If both expressions are positive or negative, then we can apply intersection of both inlist expressions
+///     1. `a in (1,2,3) AND a in (4,5) -> a in (1,2,3,4,5)`
+///     2. `a not int (1,2,3) AND a not in (4,5,6) -> a not in (1,2,3,4,5,6)`
+/// If one of the expressions is negated, then we apply exception on positive expression
+///     1. `a in (1,2,3) AND a not in (1,2,3,4,5) -> a in ()`
+///     2. `a not in (1,2,3) AND a in (1,2,3,4,5) -> a in (4,5)`
+pub(super) struct InListSimplifier {}
+
+impl InListSimplifier {
+    pub(super) fn new() -> Self {
+        Self {}
+    }
+}
+
+impl TreeNodeRewriter for InListSimplifier {
+    type N = Expr;
+
+    fn mutate(&mut self, expr: Expr) -> Result<Expr> {
+        if let Expr::BinaryExpr(BinaryExpr { left, op, right }) = &expr {
+            if let (Expr::InList(l1), Operator::And, Expr::InList(l2)) =
+                (left.as_ref(), op, right.as_ref())
+            {
+                if l1.expr == l2.expr && !l1.negated && !l2.negated {
+                    return inlist_intersection(l1, l2, false);
+                } else if l1.expr == l2.expr && l1.negated && l2.negated {
+                    return inlist_intersection(l1, l2, true);
+                } else if l1.expr == l2.expr && !l1.negated && l2.negated {
+                    return inlist_except(l1, l2);
+                } else if l1.expr == l2.expr && l1.negated && !l2.negated {
+                    return inlist_except(l2, l1);
+                }
+            }
+        }
+
+        Ok(expr)
+    }
+}
+
+fn inlist_intersection(l1: &InList, l2: &InList, negated: bool) -> Result<Expr> {
```

Comment:
I think you could avoid a clone here potentially because the mutate function gets passed an `owed` expr, like:

```
fn inlist_intersection(l1: InList, l2: InList, negated: bool) -> Result<Expr> {
```

(we could do this as a follow on PR too)

Reply:
But we still need `expr` or `Binaray Expr` if failed to match `if let` statement. I did not find out how to get the cloned value if we pass the value at first hand.

Reply:
I spent some time fooling with it this afternoon and here is what I came up with https://github.com/apache/arrow-datafusion/pull/8971